### PR TITLE
Daofund migration from Python to Java

### PR DESCRIPTION
**Pull request by olafpena**

### CODE REVIEW

File: daofund-score/src/main/java/com/iconbet/score/daofund/DaoFund.java [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+	/***
+	
+	def remove_array_item(array_db, target) -> bool:
+	    _out = array_db[-1]
+	    if _out == target:
+	        array_db.pop()
+	        return True
+	    for index in range(len(array_db) - 1):
+	        if array_db[index] == target:
+	            array_db[index] = _out
```
Comment (ismaelmk) [inline daofund-score/src/main/java/com/iconbet/score/daofund/DaoFund.java:58]: very confusing py logic, we can to test it python cli and convert it to java, it is not implemented yet (line 44)
anyways we need to confirm from iconbet devs if the logic is correct

File: daofund-score/src/main/java/com/iconbet/score/daofund/DaoFund.java [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+			BigInteger now = BigInteger.valueOf(Context.getBlockTimestamp());		
+			BigInteger day = BigInteger.ZERO;
+			day = day.add(now.divide(X_6));
+			
+			this.withdraw_count.set(_withdraw_count);
+			getWithdrawRecord("withdraw_amount").set(_withdraw_count.intValue(),_amount.toString() ) ;
+			getWithdrawRecord("withdraw_address").set(_withdraw_count.intValue(),_address.toString() ) ;
+			getWithdrawRecord("withdraw_memo").set(_withdraw_count.intValue(), _memo ) ;
+			getWithdrawRecord("withdraw_timestamp").set(_withdraw_count.intValue(), day.toString() ) ;
+			// self.icx.transfer(_address, _amount)
```
Comment (ismaelmk) [inline daofund-score/src/main/java/com/iconbet/score/daofund/DaoFund.java:157]: i do not see the java  Context.transfer call here

File: daofund-score/src/main/java/com/iconbet/score/daofund/DaoFund.java [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+    private static final String ADMINS = "admins";
+    private static final String WITHDRAW_COUNT = "withdraw_count";
+    private static final String WITHDRAW_RECORD = "withdraw_record";
+    private static final BigInteger X_6 = new BigInteger("1000000"); // 10 ** 6
+    private static final BigInteger BATCH_SIZE = BigInteger.valueOf(100L);
+    
+
+	private final ArrayDB<Address> admins = Context.newArrayDB(ADMINS, Address.class);
+	private final VarDB<BigInteger> withdraw_count = Context.newVarDB(WITHDRAW_COUNT, BigInteger.class);
+	private final DictDB<String,LinkedList> withdraw_record = Context.newDictDB(WITHDRAW_RECORD, LinkedList.class);
```
Comment (ismaelmk) [inline daofund-score/src/main/java/com/iconbet/score/daofund/DaoFund.java:30]: probably we will need to use BranchDB
https://github.com/cacomixtle/iconbet-java/blob/master/reward-distribution/src/main/java/com/iconbet/score/reward/RewardDistribution.java#L56

pretty similar the way it is accessed in py

self.withdraw_record[_withdraw_count]['withdraw_memo'] = _memo

py
variable[int index][String index] 

java
this._wagers.at(index).get(address);

https://github.com/cacomixtle/iconbet-java/blob/master/reward-distribution/src/main/java/com/iconbet/score/reward/RewardDistribution.java#L261

we will know once we deploy to dev :(

File: daofund-score/src/main/java/com/iconbet/score/daofund/DaoFund.java [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+	}
+		 
+	@External
+	public void withdraw_fund( Address _address, BigInteger _amount, String _memo) {
+		Address sender = Context.getCaller();
+		
+		if (!containsInArrayDb(sender, this.admins)) {
+			Context.revert(TAG + ": Only admins can run this method.");
+		}
+		//  _available_amount = self.icx.get_balance(self.address)
```
Comment (ismaelmk) [inline daofund-score/src/main/java/com/iconbet/score/daofund/DaoFund.java:137]: i do not see Context.getBalance java call

File: daofund-score/src/main/java/com/iconbet/score/daofund/DaoFund.java (hunk 2 above)
Comment (olafpena) [reply daofund-score/src/main/java/com/iconbet/score/daofund/DaoFund.java:162]: done.

File: daofund-score/src/main/java/com/iconbet/score/daofund/DaoFund.java (hunk 3 above)
Comment (olafpena) [reply daofund-score/src/main/java/com/iconbet/score/daofund/DaoFund.java]: Updated accordingly.

File: daofund-score/src/main/java/com/iconbet/score/daofund/DaoFund.java (hunk 4 above)
Comment (olafpena) [reply daofund-score/src/main/java/com/iconbet/score/daofund/DaoFund.java]: done.

File: daofund-score/src/main/java/com/iconbet/score/daofund/DaoFund.java (hunk 1 above)
Comment (olafpena) [reply daofund-score/src/main/java/com/iconbet/score/daofund/DaoFund.java]: Added missing method.

**Pull request merged**